services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,453 @@
+    return {expires: taskcluster.fromNow('96 hours')};
+  }
+
+  async scanPrepare() {
+    this.seen = {};
+  }
+
+  async checkWorker({worker}) {
+    const states = this.Worker.states;
+    this.seen[worker.workerPoolId] = this.seen[worker.workerPoolId] || 0;
+    if (worker.providerData.registrationExpiry &&
+      worker.state === states.REQUESTED &&
+      worker.providerData.registrationExpiry < Date.now()) {
+      return await this.removeWorker({worker});
+    }
+
+    let state = worker.state;
+    try {
+      const {provisioningState} = await this._enqueue('get', () => this.computeClient.virtualMachines.get(
+        worker.providerData.resourceGroupName,
+        worker.providerData.vm.name,
+      ));
+      if (['Creating', 'Updating', 'Starting', 'Running', 'Succeeded'].includes(provisioningState)) {
+        this.seen[worker.workerPoolId] += worker.capacity || 1;
+
+        // If the worker will be expired soon but it still exists,
+        // update it to stick around a while longer. If this doesn't happen,
+        // long-lived instances become orphaned from the provider. We don't update
+        // this on every loop just to avoid the extra work when not needed
+        if (worker.expires < taskcluster.fromNow('1 day')) {
+          await worker.modify(w => {
+            w.expires = taskcluster.fromNow('1 week');
+          });
+        }
+      } else if (['Failed', 'Canceled', 'Deleting', 'Deallocating', 'Stopped', 'Deallocated'].includes(provisioningState)) {
+        await this.removeWorker({worker});
+        this.monitor.log.workerStopped({
+          workerPoolId: worker.workerPoolId,
+          providerId: this.providerId,
+          workerId: worker.workerId,
+        });
+        state = states.STOPPED;
+      } else {
+        await worker.workerPool.reportError({
+          kind: 'creation-error',
+          title: 'Encountered unknown VM provisioningState',
+          description: `Unknown provisioningState ${provisioningState}`,
+        });
+      }
+    } catch (err) {
+      if (err.code !== 'ResourceNotFound') {
+        throw err;
+      }
+      await this.removeWorker({worker});
+      this.monitor.log.workerStopped({
+        workerPoolId: worker.workerPoolId,
+        providerId: this.providerId,
+        workerId: worker.workerId,
+      });
+      state = states.STOPPED;
+    }
+    await worker.modify(w => {
+      const now = new Date();
+      if (w.state !== state) {
+        w.lastModified = now;
+      }
+      w.lastChecked = now;
+      w.state = state;
+    });
+  }
+
+  async scanCleanup() {
+    this.monitor.log.scanSeen({providerId: this.providerId, seen: this.seen});
+    await Promise.all(Object.entries(this.seen).map(async ([workerPoolId, seen]) => {
+      const workerPool = await this.WorkerPool.load({
+        workerPoolId,
+      }, true);
+
+      if (!workerPool) {
+        return; // In this case, the workertype has been deleted so we can just move on
+      }
+    }));
```

Comment:
Is this copy-pasta from somewhere? It seems like an unnecessary thing to do.

Reply:
Yeah, this is pasta from the google provider. I wasn't totally sure of the purpose there, but we send a monitor message for `scanSeen` there, this replicates that